File: pdf_security_cleaner.rb
#!/usr/bin/env ruby

require 'hexapdf'
require 'logger'
require 'fileutils'

# Crear un logger global
$logger = Logger.new(STDOUT)
$logger.formatter = proc do |severity, datetime, progname, msg|
  "[#{datetime.strftime('%Y-%m-%d %H:%M:%S')}] #{msg}\n"
end

def deep_clean_pdf(input_file, output_file)
  $logger.info("Limpieza profunda de #{input_file}")
  doc = HexaPDF::Document.open(input_file)
  doc.pages.each do |page|
    deep_clean_page(page)
  end
  clean_document_catalog(doc)
  doc.write(output_file, optimize: true)
  $logger.info("PDF limpio guardado como #{output_file}")
end

def deep_clean_page(page)
  page.delete(:Annots)
  page.delete(:AA)
  page.delete(:JS)
end

def clean_document_catalog(doc)
  root = doc.catalog
  keys_to_remove = [:Names, :OpenAction, :AA, :AcroForm, :Outlines, :JavaScript]
  keys_to_remove.each do |key|
    root.delete(key)
  end
end

def sanitize_with_hexapdf(input_file, output_file)
  $logger.info("Sanitizando #{input_file} con HexaPDF")
  begin
    doc = HexaPDF::Document.open(input_file)
    
    # Desactivar compresión de objetos
    doc.config['object_streams.write'] = false
    
    # Linearizar el documento
    doc.config['linearize'] = true
    
    # Optimizar el documento
    doc.config['optimize_objects'] = true
    
    # Escribir el documento optimizado
    doc.write(output_file, optimize: true)
    
    $logger.info("Sanitización con HexaPDF completada")
    true
  rescue => e
    $logger.error("Error al sanitizar con HexaPDF: #{e.message}")
    false
  end
end

def clean_pdf(input_file)
  base_name = File.basename(input_file, ".*")
  temp_file1 = "#{base_name}_temp1.pdf"
  temp_file2 = "#{base_name}_temp2.pdf"
  output_file = "#{base_name}_limpio.pdf"

  # Paso 1: Limpieza profunda con HexaPDF
  deep_clean_pdf(input_file, temp_file1)

  # Paso 2: Sanitizar con HexaPDF
  if sanitize_with_hexapdf(temp_file1, temp_file2)
    # Paso 3: Segunda pasada de limpieza profunda
    deep_clean_pdf(temp_file2, output_file)
    File.delete(temp_file1)
    File.delete(temp_file2)
  else
    $logger.warn("Usando el resultado de la primera limpieza como salida final")
    FileUtils.mv(temp_file1, output_file)
    File.delete(temp_file2) if File.exist?(temp_file2)
  end

  $logger.info("Proceso de limpieza completado")
end

if ARGV.empty?
  puts "Uso: #{$PROGRAM_NAME} archivo1.pdf [archivo2.pdf ...]"
  exit 1
end

ARGV.each do |file|
  if File.file?(file) && File.extname(file).downcase == '.pdf'
    clean_pdf(file)
  elsif !File.exist?(file)
    $logger.error("Error: El archivo #{file} no existe.")
  else
    $logger.error("Error: #{file} no es un archivo PDF.")
  end
end

File: pdf_security_cleaner.py
#!/usr/bin/env python3

import os
import sys
import subprocess
import logging
from PyPDF2 import PdfReader, PdfWriter

logging.basicConfig(level=logging.INFO, format='[%(asctime)s] %(message)s', datefmt='%Y-%m-%d %H:%M:%S')

def remove_js_and_aa(input_file, output_file):
    logging.info(f"Procesando {input_file}")
    
    reader = PdfReader(input_file)
    writer = PdfWriter()

    for page in reader.pages:
        writer.add_page(page)

    # Eliminar JavaScript y Acciones Automáticas
    writer.remove_links()
    for page in writer.pages:
        if "/AA" in page:
            del page["/AA"]
        if "/Annots" in page:
            annotations = page["/Annots"]
            for annotation in annotations:
                if "/A" in annotation:
                    del annotation["/A"]

    # Eliminar JavaScript y Acciones del catálogo del documento
    if "/Names" in writer._root_object:
        del writer._root_object["/Names"]
    if "/OpenAction" in writer._root_object:
        del writer._root_object["/OpenAction"]
    if "/AA" in writer._root_object:
        del writer._root_object["/AA"]

    with open(output_file, 'wb') as f:
        writer.write(f)

    logging.info(f"PDF limpio guardado como {output_file}")

def sanitize_with_qpdf(input_file, output_file):
    logging.info(f"Sanitizando {input_file} con qpdf")
    try:
        subprocess.run([
            "qpdf", "--linearize", "--object-streams=disable",
            "--remove-unreferenced-resources=yes",
            "--qdf", "--reformat-objects", "--compress-streams=n",
            input_file, output_file
        ], check=True, capture_output=True, text=True)
        logging.info("Sanitización con qpdf completada")
    except subprocess.CalledProcessError as e:
        logging.error(f"Error al sanitizar con qpdf: {e.stderr}")
        return False
    return True

def clean_pdf(input_file):
    base_name = os.path.splitext(input_file)[0]
    temp_file = f"{base_name}_temp.pdf"
    output_file = f"{base_name}_limpio.pdf"

    # Paso 1: Remover JS y AA con PyPDF2
    remove_js_and_aa(input_file, temp_file)

    # Paso 2: Sanitizar con qpdf
    if sanitize_with_qpdf(temp_file, output_file):
        os.remove(temp_file)
    else:
        logging.warning("Usando el resultado de PyPDF2 como salida final")
        os.rename(temp_file, output_file)

    logging.info("Proceso de limpieza completado")

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(f"Uso: {sys.argv[0]} archivo1.pdf [archivo2.pdf ...]")
        sys.exit(1)

    for file in sys.argv[1:]:
        if os.path.isfile(file) and file.lower().endswith('.pdf'):
            clean_pdf(file)
        elif not os.path.isfile(file):
            logging.error(f"Error: El archivo {file} no existe.")
        else:
            logging.error(f"Error: {file} no es un archivo PDF.")